WhateverGreen/kern_igfx_memory.cpp

```diff
@@ -0,0 +1,199 @@
+//
+//  kern_igfx_memory.cpp
+//  WhateverGreen
+//
+//  Created by FireWolf on 8/30/20.
+//  Copyright © 2020 vit9696. All rights reserved.
+//
+
+#include "kern_igfx.hpp"
+#include <Headers/kern_util.hpp>
+#include <Headers/kern_disasm.hpp>
+
+void IGFX::DVMTCalcFix::init() {
+	// We only need to patch the framebuffer driver
+	requiresPatchingGraphics = false;
+	requiresPatchingFramebuffer = true;
+}
+
+void IGFX::DVMTCalcFix::deinit() {}
+
+void IGFX::DVMTCalcFix::processKernel(KernelPatcher &patcher, DeviceInfo *info) {
+	// Guard: Disable the patch if it is not available on the current Intel platforms
+	if (!available) {
+		SYSLOG("igfx", "DVMT: This fix is not available on the current platform and has been disabled.");
+		return;
+	}
+		
+	// Enable the fix if designated boot argument or device property is found
+	enabled = checkKernelArgument("-igfxdvmt");
+	if (!enabled)
+		enabled = info->videoBuiltin->getProperty("enable-dvmt-calc-fix") != nullptr;
+	
+	// Read the DVMT preallocated memory set in BIOS from the GMCH Graphics Control field at 0x50 (PCI0,2,0)
+	// TODO: Lilu needs to be updated to define the enumeration case `kIOPCIConfigGraphicsControl`
+	auto gms = WIOKit::readPCIConfigValue(info->videoBuiltin, /*WIOKit::kIOPCIConfigGraphicsControl*/ 0x50, 0, 16) >> 8;
```

Comment:
To read PCI configuration value you need call `WIOKit::awaitPublishing` as it will otherwise crash on Big Sur.

Reply:
Fixed. Thanks for pointing that.